File: prisma/refs/copy-linked-list-with-random-pointer.ts
type RandomNode = { val: number; next: RandomNode | null; random: RandomNode | null };

export function copyRandomList(head: RandomNode | null): RandomNode | null {
  const map = new Map<RandomNode, RandomNode>();
  for (let cur = head; cur; cur = cur.next) {
    map.set(cur, { val: cur.val, next: null, random: null });
  }
  for (let cur = head; cur; cur = cur.next) {
    const copy = map.get(cur)!;
    copy.next = cur.next ? map.get(cur.next)! : null;
    copy.random = cur.random ? map.get(cur.random)! : null;
  }
  return head ? map.get(head)! : null;
}

File: prisma/refs/py/copy-linked-list-with-random-pointer.py
def copy_random_list(head):
    mapping = {}
    cur = head
    while cur:
        mapping[id(cur)] = Node(cur.val)  # noqa: F821 (injected by harness)
        cur = cur.next
    cur = head
    while cur:
        copy = mapping[id(cur)]
        copy.next = mapping[id(cur.next)] if cur.next else None
        copy.random = mapping[id(cur.random)] if cur.random else None
        cur = cur.next
    return mapping[id(head)] if head else None
